fastlane/Fastfile

```diff
@@ -1,49 +1,107 @@
-# You can find the documentation at https://docs.fastlane.tools
-#
-# For a list of all available actions, check out
-#
-#     https://docs.fastlane.tools/actions
-#
-# For a list of all available plugins, check out
-#
-#     https://docs.fastlane.tools/plugins/available-plugins
-#
-
-# Uncomment the line if you want fastlane to automatically update itself
-# update_fastlane
+require_relative './amplify_pods.rb'
 
 opt_out_usage
-
 default_platform(:ios)
 
-COCOAPODS_PUSH_WAIT_TIME = 600
+PLIST_KEY = "CFBundleShortVersionString"
 
-pods = [
-  {spec: "Amplify.podspec", sleep: true},
-  # AmplifyPlugins depends on AWSPluginsCore, but skipping the sleep here
-  # since there's one after CoreMLPredictionsPlugin
-  {spec: "AWSPluginsCore.podspec", sleep: false},
-  {spec: "CoreMLPredictionsPlugin.podspec", sleep: true},
-  {spec: "AWSPredictionsPlugin.podspec", sleep: false},
-  {spec: "AmplifyPlugins.podspec", sleep: false}
-]
+COCOAPODS_PUSH_WAIT_TIME = 600
 
 platform :ios do
+
+  desc "Create a pre-release version by pushing a tag to GitHub, and pushing pods to CocoaPods"
+  lane :unstable do 
+
+    next_version = calculate_next_canary_version
+
+    UI.message("Release unstable version: #{next_version}")
+
+    # Increment all specs and plists
+    increment_versions(version: next_version)
+
+    # Create tag and push to origin
+    add_tag(version: next_version)
+
+    # Push to CocoaPods
+    release_pods()
+  end
+
+  desc "Create a release version by building and committing a changelog, pushing a tag to GitHub, and pushing pods to CocoaPods"  
+  lane :release do
+
+    next_version, commits = calculate_next_release_version
+
+    UI.message("Release version: #{next_version}")
+
+    # Increment all specs and plists
+    increment_versions(version: next_version)
+
+    changelog = build_changelog(version: next_version, commits: commits)
+
+    write_changelog(changelog: changelog, path: 'CHANGELOG.md')
+
+    # Commit and push
+    release_commit(version: next_version)
+
+    # Create tag and push to origin
+    add_tag(version: next_version)
+
+    # Push pods
+    release_pods()
+  end
+
+  desc "Increment versions"
+  private_lane :increment_versions do |options|
+    version = options[:version].to_s
+
+    AmplifyPods.pods.each do |pod|
+      spec, push, constants, plist_paths = pod.values
+
+      UI.message("Incrementing version for #{spec}")
+
+      constants.each do |constant|
+        set_key_value(file: spec, key: constant, value: version)
+      end
+
+      plist_paths.each do |plist_path|
+        set_info_plist_value(path: plist_path, key: PLIST_KEY, value: version)
+      end
+    end
+  end
+  
+  desc "Commit and push"
+  private_lane :release_commit do |options|
+    next_version = options[:version]
+
+    sh("git config --global user.email $GITHUB_EMAIL")
+    sh("git config --global user.name $GITHUB_USER")
+
+    commit_message = "chore(release): #{next_version} [skip ci]"
```

Comment:
Just Bikeshedding on a Friday: Why not just `release: `? Not sure the chore type buys us much, and release is a weird scope.

Reply:
`release:` is technically not a valid type (to the best of my knowledge). I went with what we have on JS but can change it to e.g., `chore: release  #{next_version} [skip ci]`

Reply:
There's actually nothing in the specification that prevents you from using your own nouns for types. The only defined types are `fix` and `feat`. The [specification](https://www.conventionalcommits.org/en/v1.0.0/#specification):

> 4. types other than fix: and feat: are allowed, for example @commitlint/config-conventional (based on the the Angular convention) recommends build:, chore:, ci:, docs:, style:, refactor:, perf:, test:, and others.

Commitlint apparently has a default list, but we should feel like we can expand upon this. The specification only concerns itself with `feat` and `fix`.

Reply:
That's fair, I should've phrased that differently. It's not invalid per se, just not part of the convention

Reply:
I'm being a pedant here, but bear with me. The scope is defined as:

> 4. A scope MAY be provided after a type. A scope MUST consist of a noun describing a section of the codebase surrounded by parenthesis, e.g., fix(parser):

Release isn't technically a scope by the above definition. It seems like more of a type. I'm surprised this isn't more commonly used.

Reply:
I agree with your point that `release` is nonsensical as a scope. So I'm suggesting dropping the scope and putting "release" in the description, following a **chore** type.

E.g., `chore: release #{next_version} [skip ci]`